File: trunk/Project/Test/src/Princess/Character.java

/**
    キャラクター情報を保持するクラス.
*/
class Character
{
    String m_name;
    int m_hp;
    int m_hp_max;
    int m_attack;
    int m_exp; //!< 経験値.
    final int m_exp_max = 10 ;
    final float m_grow_persent = 1.7f; //!< 成長係数.
    
    Character( String name , int hp_max , int attack , int exp )
    {
        m_name = name ;
        m_hp_max = hp_max ;
        m_attack = attack ;
        m_exp = exp;
        
        FullRecovery();
    }
    
    void FullRecovery()
    {
        m_hp = m_hp_max;
    }
    
    void LevelUp()
    {
        while( m_exp >= m_exp_max )
        {
            m_hp_max *= m_grow_persent;
            m_attack *= m_grow_persent;
            m_exp -= m_exp_max;
            FullRecovery();
        }
    }
};

File: trunk_dxlib/Source/Project/Character.cpp
#include "Project/Character.hpp"
#include "DxLibWrapper/Random.hpp"

namespace{

static int const s_default_exp_max = 100;

}//namespace unnamed

Character::Character( int hp_max , int attack , int exp )
 : m_hp_max( hp_max )
 , m_hp( hp_max )
 , m_attack( attack )
 , m_exp( exp )
 , m_exp_max( s_default_exp_max )
{
    m_select_index = 0 ;
}

bool Character::IsLive() const
{
    return ( m_hp > 0 );
}

int Character::AttackDamage() const
{
    float const rand = static_cast<float>( GetRandToMax( 4 ) );
    float keisuu = ( rand - 2 ) / 10 ;
    
    return static_cast<int>( m_attack + m_attack * keisuu );
}


